### ISSUE TITLE
Ability to down-sample time-series data

### ISSUE THREAD

**Issue by jarohen** [xtdb/xtdb#3515]
Beyond a certain age, it's unnecessary to store time-series data at its original resolution. For queries over longer periods of valid-time, we'd prefer not to have to scan data at this resolution - a down-sampled aggregation is plenty sufficient.

e.g.:
* We may only be interested in 5-min granularity data for a few months
* Queries summarising the past year may be fine with daily readings
* etc etc.

Our normal temporal filters may not help enough here - even though we can (#3166) filter down to only the temporally relevant data relatively well, there's still potentially a lot of temporally relevant data to scan through.

(related, out of scope of this card: we may also consider later adding retention periods - periods beyond which we no longer retain data at the original granularity. e.g. after a year, because we've rolled up the 5-min data into daily readings, we can ~safely erase older 5-min readings. This is less urgent, because this data should naturally fall out of node-local caches, and object-storage is considered relatively cheap)

Prior art:

* Timescale ['continuous aggregates'](https://docs.timescale.com/use-timescale/latest/continuous-aggregates/) - these support a subset of SQL
* Influx [down-sampling](https://docs.influxdata.com/influxdb/v2/process-data/common-tasks/downsample-data/) - expressed as tasks that run periodically.
* Postgres (et al) [materialized views](https://www.postgresql.org/docs/current/rules-materializedviews.html) - these are manually refreshed by the user.
* [Differential dataflow](https://timelydataflow.github.io/differential-dataflow/) - this is significantly more general than what's required for down-sampling individual entities' histories, but a good one to keep in mind if we want to extend into this direction in the future.
* [Lambda architecture](https://en.wikipedia.org/wiki/Lambda_architecture) - a recommendation to split systems into three: a batch layer that regularly re-computes the bulk of the view, a speed layer to give access to the most recent data, and a serving layer to bring them together.

## API

For the down-sampling above, the SQL we'd need to support would be something akin to the following:

```sql
-- exact name tbc
CREATE (MATERIALIZED|CONTINUOUS|INCREMENTAL) VIEW readings_daily AS
  WITH bucketed_readings AS (
    SELECT _id, DATE_TRUNC('DAY', _valid_from) AS day, reading
    FROM readings_raw FOR ALL VALID_TIME
  )

  SELECT _id, day AS _valid_from, day + INTERVAL 'PT1D' AS _valid_to,
         SUM(reading) AS sum, COUNT(reading) AS count
  FROM bucketed_readings
  GROUP BY _id, day
```

This would then get planned into a query with three distinct operators - a scan, a projection (for the date-trunc), and an aggregation. Eventually we'd likely want to support the other relational operators, but, for now, these should be a good start.

I did alternatively consider a bespoke syntax for down-sampling, likely a lot more constrained and specific to this use-case (e.g. specifying the source table, destination table, window and aggregation functions), but on balance I'd prefer to stick to SQL if possible.

## Initial implementation approach

(miscellaneous unsorted thoughts)

* In the differential data-flow algebra, the intermediate results are similar to standard relations, except that rows (deltas) also carry a 'multiplicity' - this reflects whether (/how many times) the row was added or deleted; unwinding this relation at the end of the delta query yields the output.
* Bitemporality obviously makes this more fun :sweat_smile: 
* XT2 already has a compaction process for its primary index, compacting chunks into deeper levels of its LSM tree. These take the form of put/delete events - system-from, valid-from, valid-to, id and document. At scan time, we then perform bitemporal resolution to understand which rows are temporally relevant to the query at hand.
* These events are already quite useful for delta tables, in their raw form - hopefully we can make use of these without needing to transform/duplicate them. We're also required to keep track of the rows that are either replaced or deleted, which we'll have to maintain separately - we'll need to consider how to make this efficient.
* It'd be useful if the logical output of this step was a standard bitemporal relation, so that it could then fit straight into other queries (as a standard SQL view would).

## Milestones

- [ ] Keeping track of what views have been requested - parsing the above DDL, maintaining a set of queries.
- [ ] Compactor recognises what work is available
- [ ] Incremental scan operator - maintaining a log of bitemporal retractions caused by a file.
- [ ] Incremental projection operator - given scan deltas, output projection deltas
- [ ] Incremental aggregation operator
- [ ] (Real) scan operator pulls in data from delta views, and combines it with data that hasn't yet made it into a compacted chunk to give up-to-date results.

**Comment by jarohen** [comment]
### Chalk — What a materialised view means in a bitemporal database

**tl;dr** — the semantics of a materialised view in a bitemporal database, settled far enough to scope an implementation. Design only; no code.

- **An MV is a function from system time to a relation**
  `MV : SystemTime -> Relation` — the same kind of object a base table already is. System time stays a parameter; valid time lives in the data.

- **Which makes `FOR SYSTEM_TIME AS OF t` the only temporal operator an MV can offer, and it is enough**
  `FOR ALL SYSTEM_TIME` is out, because its answer is not derivable in a size proportional to itself.

- **XTDB's LSM has one key, so the line between cheap and expensive is single-table versus join**
  A bucket rollup keys on the base `_id` and needs no new storage concept. A join needs its inputs arranged by the join key, which is a secondary index that does not exist.

- **Downsampling does not need the general machinery — but corrections force bitemporal resolution regardless**
  Nothing in the log writes "minus the old reading", so resolution has to manufacture the retraction half of every correction before any aggregate can see it.

- **Unfinished:** cumulative sums, and whether broadcast joins bring the expensive `fusion.sql` grid queries within reach of a first version.

- [x] P1/P2 — two core temporal principles, stated as constraints
- [x] Separating the two temporal axes
- [x] What an MV is: `SystemTime -> Relation`
- [x] Which temporal operators an MV can offer
- [x] Definitions are immutable
- [x] Data shapes mapped to view shapes
- [x] Where the storage model draws the line
- [x] Corrections work from day one
- [ ] Q1 — cumulative sums, not reached
- [ ] Q2 — broadcast-join escape hatch, cardinalities not measured

<details><summary>P1/P2 — two core temporal principles, stated as constraints</summary>

Everything below was tested against these two. Cited by ID elsewhere in this comment.

- **P1 — an MV's result MUST be correct and consistent as at the basis of the wider query**
  A cumulative sum returned by an MV cannot be inconsistent with anything else the same read-transaction returns.

- **P2 — if a user could reasonably make a decision off an MV's result, they MUST be able to go back in system time and query what that MV *would have* returned**
  This is what rules out an MV that only holds current state and silently answers a past basis from it.

- **Together they give the contract: as-of-now queries are fast, historical queries are possible**
  Refusing a historical query is not an option, and the query text does not change between the two cases.

</details>

<details><summary>Separating the two temporal axes</summary>

Splitting the two axes is what turns one hard problem into a non-problem and a real problem.

- **System time is the circuit's own time axis; valid time is a data dimension**
  Almost everything else in this comment follows from taking that split seriously.

- **`FOR SYSTEM_TIME AS OF NOW` is free**
  Nothing can exist above the durable head, so wall-clock-now and the query basis are indistinguishable — there is no observable nondeterminism to solve.

- **`FOR VALID_TIME AS OF NOW` is the hard one, because the valid axis has a fully populated future and the system axis has none**
  Facts with `_valid_to` next Tuesday exist in the store right now, so the answer changes with no input delta at all. That asymmetry is the entire difference between the two axes, and the reason one word needs two implementations.

- **No wall clock goes into the circuit, ever — needed twice over**
  Replica consistency (two replicas must compute the same view from the same log), and deterministic replay from a checkpoint.

  - **Taking the tick from the log instead would be deterministic but wrong**
    A contract expiring at midnight would not leave the view until the next transaction arrived.

- **So the clock-driven part is necessarily a read-time concern, for a structural reason rather than a preference**
  Time passing has no `_system_from`. It is not a transaction, so there is nothing to stamp it with, so it cannot be recorded in a bitemporal table at all.

</details>

<details><summary>What an MV is: `SystemTime -> Relation`</summary>

System time stays an input; valid time is part of the data.

- **That is the same kind of object a base table already is**
  `readings` at basis B is a relation whose rows carry valid-time columns. Which retires the "is an MV a table?" question — it is the same *function* a table is, independent of how it is stored.

- **P1 and P2 then hold by construction**
  The wider query's basis parameterises the MV, so consistency is automatic; MV(t) is defined for every t, so P2 is definitional and only *speed* is ever at stake.

- **`FOR VALID_TIME` clauses become ordinary predicates on data columns**
  No propagation mechanism needed.

- **Which makes the two useful cases one rule with different columns**
  Bucket bounds in `_valid_from`/`_valid_to` gives "the rollup in force at T"; the bucket as a plain column with `_valid_from` = computation time gives "the rollup table as it stood at T". Both worth carrying.

- **The DDL sketch already in this card is the first of those, and it holds up**
  `_id` plus `day AS _valid_from` / `day + INTERVAL 'PT1D' AS _valid_to` is exactly the shape reached from the other direction, and it is what lets a rollup be stored as an ordinary table rather than needing a new storage concept.

</details>

<details><summary>Which temporal operators an MV can offer</summary>

Decomposed by operator, and the set is exhaustive over the temporal syntax.

- **`FOR SYSTEM_TIME AS OF t` — yes, and it is total**
  It is MV(t), always recomputable from the definition plus the base tables.

- **`FOR ALL SYSTEM_TIME` and `BETWEEN` — no, and not on taste**
  The answer is not derivable in a size proportional to itself. A base table stores each row's system-time extent, so the query is O(answer); an MV would have to evaluate at every transaction and diff, which is O(history). A complexity-class difference, not a constant.

  - **And supporting it would make the cache semantically load-bearing**
    The answer would exist only where something happened to be materialised, which is the one thing an MV must never do.

- **`FOR VALID_TIME` in any form — yes, as a data predicate**
  Available exactly when the definition projects valid-time columns, which is a property of the schema rather than a language restriction.

- **The rule that falls out: an MV supports exactly the temporal operations that recompute can serve**
  Nothing else then needs deciding case by case.

</details>

<details><summary>Definitions are immutable</summary>

A definition change means a new view; the old one is kept as long as its logic needs defending.

- **Version collapses into identity**
  No `definition_version` field on checkpoints, no fork-or-invalidate logic, no "which logic produced this row". The view's name *is* its version.

- **Which makes "as reported in March" both answerable and cheap**
  The March report was produced by a view that still exists, so there is no old logic to reconstruct.

- **The strongest argument for it is honesty rather than convenience**
  Mutating a definition retroactively changes what past reports meant, leaving the old numbers attributable to no logic that exists. Immutability makes that inexpressible.

- **The cost lands on consumers, who have to be repointed at the new name**
  See Q5.

</details>

<details><summary>Data shapes mapped to view shapes</summary>

Worked from the shapes in `dev/doc/evaluating-index-strategies.adoc` against the queries in `modules/bench/src/main/resources/xtdb/bench/fusion.sql` and `tsbs.sql`. Decomposed one child per data shape in that doc.

- **Four recurring view types, after dropping one**
  Bucketed rollup; denormalised temporal join; cumulative running total; grid rollup with retraction.

- **Last-value-per-entity was dropped as an MV shape**
  The scan already makes it fast — it is priority 1 in the index-strategies doc and what the trie design is *for* — so nobody would create a view for it. Pre-resolving the current plane during compaction may still pay on superseded-wide tables, but that is a trie-layout decision, not a user-facing view.

- **The mapping is sparse, and the sparsity has a cause**
  Append-only facts attract bucket rollups; settle-and-stay entities attract denormalised joins and grid rollups; anything carrying a signed running quantity attracts cumulative totals. The temporal character of the data determines the view type, so the cross product was never going to fill in.

  - **IoT readings and pricing** — bucket rollups only.

  - **User profiles** — denormalised read models, where the cost driver is version fan-out (a view joined to six dimensions re-versions whenever any of them changes), not aggregate state.

  - **Trades and orders** — cumulative totals and grid rollups, plus risk as join-then-aggregate. The hardest shape and the most valuable.

  - **Static reference data** — no views of its own; it is the small side of everyone else's joins, and the machinery should recognise that and build nothing.

- **The expensive queries in `fusion.sql` are grid queries, and a grid query is the clock replayed deterministically**
  `generate_series` over an hourly grid joined to six or seven tables on `_valid_time CONTAINS gen.t`. Consecutive grid points differ by exactly the rows whose validity boundary falls between them — the same delta as a clock tick, but driven by the query, with no wall clock involved.

</details>

<details><summary>Where the storage model draws the line</summary>

The finding that most changed the shape of the plan: the discriminator is the *key*, not the aggregation complexity. XTDB's LSM is a hash trie on `_iid` — one index, one key.

- **A bucket rollup needs no other key**
  The bucket is `date_bin(interval, _valid_from)`, derivable from the incoming row with no state read at all; the output is keyed by the base `_id`; and the working set is one open bucket per entity, which lives in the live index. No LSM reads on the hot path, and no new structures.

- **A join needs its inputs arranged by the *join key*, which is not `_iid`**
  That is a secondary index XTDB does not have, with its own compaction and its own pruning metadata, sized O(table) rather than O(view).

  - **Permanent residency**
    A join must retain the full integral of both inputs, so a new row on either side can probe the other.

  - **Consolidation as a second source of truth**
    The arrangement wants collapse-to-current, and the base tries cannot do that — so it is a separate structure to keep in step.

- **Cumulative sums land on the good side**
  `sum` per entity over valid time is keyed by `_id` in the trie's own order, so it needs no secondary index. Its difficulty — a retroactive write invalidating O(suffix) — is algorithmic, not storage-model.

- **So the natural first boundary is single-table versus join, not simple versus general**
  Single-table MVs of any aggregate complexity need no new storage concept.

- **idea: `I1` — one framework staged by operator set, rather than two mechanisms**
  Proposed in session and not yet agreed. The monoid/group split is not a rival framework to DBSP: Differential Dataflow's generic `reduce` re-reduces a key's multiset and special-cases `count`/`sum` as incremental, so "which aggregates combine versus re-reduce" is a per-aggregate property you answer either way. That suggests building the keyed reduce, accepting only the definitions it can serve, and letting joins arrive later as new operators rather than as a second subsystem — with the user-facing `CREATE MATERIALIZED VIEW` surface unchanged throughout.

</details>

<details><summary>Corrections work from day one</summary>

Which means invertibility is not optional, and the `min`/`max` escape hatch belongs in the first version rather than a later one.

- **Resolution is a mandatory first stage, for a reason that does not come from wanting aggregates to be fast**
  The *retract* half of a correction does not exist in the log — nothing writes "minus the old reading". Resolution manufactures it by comparing against what was there before. Skip it and a correction double-counts into the bucket.

- **The tiling is only clean in the absence of corrections**
  `modules/datasets/src/main/clojure/xtdb/tsbs.clj:144` writes `_valid_from = timestamp` and `_valid_to = timestamp + log_interval`, so readings tile and never overlap — until a correction writes a second event for the same `(device, interval)` at a later `_system_from`. That is genuine overlap and it is `PolygonCalculator`'s job. Effective identity is user PK × chronon.

- **`sum`, `count` and `avg` are groups, so a correction is minus-old-plus-new**
  O(1), no re-scan, no circuit.

- **`min` and `max` are only monoids, but the bucket is tiny by construction**
  Retracting the current extremum needs the bucket's other values — and that is twelve rows in an hourly bucket over 5-minute readings, 288 in a daily one. The re-reduce fallback that is a genuinely bad worst case for `max(salary) GROUP BY department` is trivial here, and it is a point lookup by `(device, valid_time range)`, which is what the tries and metadata pruning are good at.

- **The distinction that matters is "replaced", not "late"**
  A late-arriving reading for a *new* interval is a pure insert, and `max(old, new)` absorbs it free. Only a replacement produces a retraction, and resolution tells you which you have.

</details>

<details><summary>Open questions</summary>

Decomposed one child per unresolved thread; IDs are stable, cite them rather than re-describing.

- **`Q1` — cumulative sums, not reached**
  Retroactive writes invalidate O(suffix), and the mitigation (deltas plus periodic checkpoints, computing forward from the nearest) converged from three separate directions in session but was never worked through. The energy client's 5-minute power-to-energy sum is the concrete case to work it against.

- **check: `Q2` — dimension cardinalities in the fusion benchmark**
  A broadcast join needs memory, not an index, and `system` / `site` / `device_model` / `organisation` are all described as low-cardinality. Hash-broadcasting them might bring the expensive fusion queries into a first version with no secondary-index subsystem at all — worth measuring before writing joins out of scope.

- **check: `Q3` — can a client stamp a *future* `system_time`?**
  `TransactionKey.system_time` is documented as non-monotonic in `dev/doc/db.allium`. If a future stamp is possible, then a `system_time as of now` read between that tx and wall clock reaching its stamp has to hide it, and the system axis acquires the same wall-clock problem as the valid axis. Decides whether "system-now is free" actually holds.

- **`Q4` — non-invertible per-key argmax outside bucket rollups**
  Three workloads independently want it: `item_max_bid` in auctionmark, tsbs's last-reading-per-truck, and the index-strategies doc's IoT line about aggregates over all entities at an instant. Top-k versus re-reduce changes the state layout, so it wants deciding before the reduce operator is written.

- **`Q5` — consumer naming under immutable definitions**
  Every consumer has to be repointed when a definition changes. A mutable alias would reintroduce the unrepeatability that immutability removed; a *bitemporal* alias resolved at the query's basis would keep repeatability and make "what did `risk_report` mean in March" answerable — but that is catalog semantics, and worth taking separately.

- **`Q6` — the performance cliff**
  Falling off the cache still returns correct results, just slowly, which is a real hazard for a dashboard that silently gets slow. Checkpoints at block boundaries turn the cliff into a slope; a cache-miss reason visible at runtime, not only in `EXPLAIN`, is the operational half of it.

</details>
